### PR TITLE
added solution

### PR DESCRIPTION
**Pull request by Mzuha**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,49 +1,100 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:7]: redundant empty line

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,49 +1,100 @@
...
 
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] storage;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:9]:
```suggestion
    private T[] storage;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,49 +1,100 @@
...
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkForAdd(index);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:44]: let's make the method name more suitable)

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,49 +1,100 @@
...
+        return (T) storage[index];
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:45]:
```suggestion
        return storage[index];
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,49 +1,100 @@
...
-
+        checkForAdd(index);
+        storage[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkForAdd(index);
+        T oldValue = (T) storage[index];
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:57]:
```suggestion
        T oldValue = storage[index];
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,49 +1,100 @@
...
+        if (size - 1 > index) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:58]: let's remove this `if`?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,49 +1,100 @@
...
+            System.arraycopy(storage, index + 1, storage, index, size - index);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:59]: let's think again about the last parameter here

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,49 +1,100 @@
...
+        T oldValue;
+        for (int i = 0; i < size; i++) {
+            if (storage[i] == element || storage[i] != null && storage[i].equals(element)) {
+                oldValue = element;
+                System.arraycopy(storage, i + 1, storage, i, size - i);
+                size--;
+                return oldValue;
+            }
+        }
+        throw new NoSuchElementException();
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:76]: add informative message to exception

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,49 +1,100 @@
...
     @Override
     public T remove(T element) {
-        return null;
+        T oldValue;
+        for (int i = 0; i < size; i++) {
+            if (storage[i] == element || storage[i] != null && storage[i].equals(element)) {
+                oldValue = element;
+                System.arraycopy(storage, i + 1, storage, i, size - i);
+                size--;
+                return oldValue;
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:73]: can we reuse `remove(int index)`?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,49 +1,100 @@
...
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void grow() {
+        int oldCapacity = storage.length;
+        int newCapacity = oldCapacity + oldCapacity / 2;
+        storage = Arrays.copyOf(storage, newCapacity);
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:92]: let's use `System.arraycopy()` instead

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Comment by Mzuha** [comment]
removed redundant lne*

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,49 +1,95 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:6]: please remove redundant empty line after class declaration

File: src/main/java/core/basesyntax/ArrayList.java [hunk 12]
```diff
@@ -1,49 +1,95 @@
...
+        storage[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:28]: should we also check capacity here?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 13]
```diff
@@ -1,49 +1,95 @@
...
+    }
+
+    private void grow() {
+        int newCapacity = storage.length + storage.length / 2;
+        T[] newStorage = (T[]) new Object[newCapacity];
+        System.arraycopy(storage, 0, newStorage, 0, size);
+        storage = newStorage;
+    }
+
+    private void indexValidation(int index) {
```
Comment (KseniiaMakarova) [inline src/main/java/core/basesyntax/ArrayList.java:90]:
```suggestion
    private void validateIndex(int index) {
```
we should use verbs for method names

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]
good job! added minor comments

**Review by KseniiaMakarova — APPROVED** [review]
great

**Pull request closed without merging**